### PR TITLE
Precision-recall curve for NoiseAwareModel

### LINKED ISSUE #466: Output precision-recall curves with score function

### PR DESCRIPTION
**Pull request by henryre**
Decided to put it in a function separate from score so that people won't accidentally threshold-hack

prcurve

Fixes #466

### CODE REVIEW

File: snorkel/learning/utils.py [hunk 1]
```diff
@@ -74,6 +74,51 @@ def score(self, test_marginals, train_marginals=None, b=0.5, set_unlabeled_as_ne
 
 class MentionScorer(Scorer):
     """Scorer for mention level assessment"""
+
+    def score_b(self, test_marginals, b, get_candidates=False, set_unlabeled_as_neg=True, set_at_thresh_as_neg=True):
```
Comment (ajratner) [inline snorkel/learning/utils.py]: I don't really get why we need to separate into `score` and `score_b`?  Maybe it's just the name in particular that bugs me.  Maybe `score_b` -> `_score_buckets`?

Comment (ajratner) [inline snorkel/learning/utils.py]: I don't really get why we need to separate into `score` and `score_b`?  Maybe it's just the name in particular that bugs me.  Maybe `score_b` -> `_score_buckets`?

File: snorkel/learning/utils.py [hunk 2]
```diff
@@ -74,6 +74,51 @@ def score(self, test_marginals, train_marginals=None, b=0.5, set_unlabeled_as_ne
...
+                        fn.add(x if get_candidates else i)
+
+        return (tp, fp, tn, fn, test_label_array)
+
+    def pr_curve(self, test_marginals, set_unlabeled_as_neg=True, set_at_thresh_as_neg=True):
+        """Display a precision-recall curve for various threshold values"""
+        bxr = np.linspace(0, 1.0, 21)
+        cts = [self.score_b(test_marginals, b, False, set_unlabeled_as_neg, set_at_thresh_as_neg)[:4] for b in bxr]
+        p, r, f1 = zip(*[scores_from_counts(*x) for x in cts])
+        with plt.style.context('fivethirtyeight'):
```
Comment (ajratner) [inline snorkel/learning/utils.py]: srsly?  That's so pre-election

**Comment by ajratner** [comment]
Also this is super nice looking and awesome

File: snorkel/learning/utils.py (hunk 1 above)
Comment (henryre) [inline snorkel/learning/utils.py]: Modularity. Reuse of code. Etc. This is called both by `pr_curve` (over multiple threshold values) and by `score`. Can rename to `_score_threshold`

File: snorkel/learning/utils.py (hunk 2 above)
Comment (henryre) [inline snorkel/learning/utils.py]: They gave it a 30% chance...

**Comment by henryre** [comment]
Agree with integrating with parameter searches. Will wrap it in

**Comment by ajratner** [comment]
@henryre Should we just merge this in without integrating with grid search, and do that later?

**Comment by ajratner** [comment]
@henryre thoughts on what to do with this?

**Comment by henryre** [comment]
@ajratner sorry just saw. Definitely still a worthwhile addition, but going to close because we're too out of date.

**Pull request closed without merging**